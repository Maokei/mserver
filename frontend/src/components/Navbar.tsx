import { useState } from 'react';
import { useNavigate } from 'react-router-dom';

const Navbar = () => {
  const navigate = useNavigate();
  const [dropDown, setDropDownState] = useState(false);
  const toggleDropDown = () => setDropDownState(!dropDown);

  const handleClick = () => {
    navigate('/');
  };

  return (
    <header className='w-full bg-white shadow p-4 sticky top-0 z-10'>
      <nav className='flex justify-between items-center'>
        <div className='text-2xl font-bold' onClick={handleClick}>
          MServer
        </div>
        <button onClick={toggleDropDown}>
          <span className='hidden md:block font-medium'>User</span>
        </button>
      </nav>
    </header>
  );
};
export default Navbar;
