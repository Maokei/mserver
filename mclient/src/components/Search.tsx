import React from "react";
import { ReactComponent as SearchIcon } from "../assets/search-vector.svg";

interface SearchProps {
    onClick: () => void;
}

export const Search: React.FC<SearchProps> = ({ onClick }) => {
    return (
        <div
            data-testid="search"
            className="control column is-four-fifths search-wrapper"
        >
            <button className="icon" onClick={onClick}>
                <SearchIcon />
            </button>
            <input className="input" type="text" placeholder="search" />
        </div>
    );
};
