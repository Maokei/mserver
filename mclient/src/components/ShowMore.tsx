import React from "react";
import { ReactComponent as MoreIcon } from "../assets/union-vector.svg";

export const ShowMore = () => {
    const ref = React.useRef<HTMLDivElement>(null);
    const [isActive, setIsActive] = React.useState<boolean>(false);
    const dropdownActiveClass = isActive ? "is-active" : "";

    React.useEffect(() => {
        const checkIfClickedOutside = (e: MouseEvent) => {
            if (
                isActive &&
                ref.current &&
                !ref.current.contains(e.target as HTMLElement)
            ) {
                setIsActive(false);
            }
        };

        document.addEventListener("mousedown", checkIfClickedOutside);

        return () => {
            document.removeEventListener("mousedown", checkIfClickedOutside);
        };
    }, [isActive]);

    return (
        <div
            className={`column dropdown is-right ${dropdownActiveClass} show-more-wrapper`}
        >
            <div className="dropdown-trigger">
                <button
                    data-testid="show-more"
                    className="button icon"
                    aria-haspopup="true"
                    aria-controls="dropdown-menu"
                    onClick={() => setIsActive(!isActive)}
                >
                    <MoreIcon />
                </button>
            </div>

            <div
                ref={ref}
                className="dropdown-menu"
                id="dropdown-menu"
                role="menu"
            >
                <div className="dropdown-content">
                    <div className="dropdown-item">Dropdown item</div>
                    <div className="dropdown-item">Other dropdown item</div>
                    <div className="dropdown-item is-active">
                        Active dropdown item
                    </div>
                </div>
            </div>
        </div>
    );
};
