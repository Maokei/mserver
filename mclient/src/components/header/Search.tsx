import React from "react";

import { SearchIcon } from "../../assets/SearchIcon";

interface SearchProps {
	onClick: () => void;
	hidden: boolean;
	value: string;
	setKeywords: any;
	onKeyDown: any;
}

export const Search: React.FC<SearchProps> = ({
	onClick,
	hidden,
	value,
	setKeywords,
	onKeyDown,
}) => {
	const hiddenClass = hidden ? "hidden" : "";

	return (
		<div
			data-testid="search"
			className="control column is-four-fifths search-wrapper"
		>
			<button
				data-testid="search-btn"
				className="button icon search-btn"
				onClick={onClick}
			>
				<SearchIcon />
			</button>
			<input
				data-testid="search-input"
				className={`input is-large ${hiddenClass}`}
				type="text"
				placeholder="search"
				value={value}
				onChange={(e: any) => setKeywords(e.target.value)}
				onKeyDown={onKeyDown}
			/>
		</div>
	);
};
