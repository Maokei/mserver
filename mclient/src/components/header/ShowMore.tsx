import React from "react";
import { Link } from "react-router-dom";
import { ShowMoreIcon } from "../../assets/union-vector";

interface ShowMoreProps {
	dropdownRef: any;
	state: boolean;
	setState: Function;
}

export const ShowMore: React.FC<ShowMoreProps> = ({
	dropdownRef,
	state,
	setState,
}) => {
	const dropdownActiveClass = state ? "is-active" : "";

	return (
		<div
			data-testid="dropdown-test"
			className={`column dropdown is-right ${dropdownActiveClass} show-more-wrapper`}
		>
			<div className="dropdown-trigger">
				<button
					data-testid="show-more"
					className="button icon"
					aria-haspopup="true"
					aria-controls="dropdown-menu"
					onClick={() => setState(!state)}
				>
					<ShowMoreIcon />
				</button>
			</div>

			<div
				data-testid="dropdown-menu-test"
				ref={dropdownRef}
				className="dropdown-menu"
				id="dropdown-menu"
				role="menu"
			>
				<div
					data-testid="dropdown-content-test"
					className="dropdown-content"
				>
					<div
						data-testid="dropdown-item-test"
						className="dropdown-item"
					>
						<Link to={"/"}>Home</Link>
					</div>
					<div
						data-testid="dropdown-item-test"
						className="dropdown-item"
					>
						<Link to={"/login"}>Login</Link>
					</div>
					<div className="dropdown-item">
						<Link to={"/signup"}>Sign Up</Link>
					</div>
				</div>
			</div>
		</div>
	);
};
