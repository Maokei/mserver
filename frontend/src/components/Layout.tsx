import { useState } from 'react';
import { Outlet } from 'react-router-dom';
import Navbar from './Navbar';
import Sidebar from './Sidebar';

const Layout = () => {
  const [isSidebarCollapsed, setIsSidebarCollapsed] = useState(false);

  const toggleSidebar = () => {
    setIsSidebarCollapsed(!isSidebarCollapsed);
  };

  return (
    <div className='h-screen flex flex-col'>
      <Navbar />
      <div className='flex flex-grow overflow-hidden'>
        {/* Sidebar */}
        <Sidebar isCollapsed={isSidebarCollapsed} toggleSidebar={toggleSidebar} />

        {/* Main Content */}
        <div
          className={`flex-grow p-4 transition-all duration-300 ${isSidebarCollapsed ? 'ml-14' : 'ml-64'}`}
        >
          <Outlet />
        </div>
      </div>
    </div>
  );
};

export default Layout;
