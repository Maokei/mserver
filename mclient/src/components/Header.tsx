import React from "react";
import { Search } from "./Search";
import { ShowMore } from "./ShowMore";

export const Header = () => {
    const [isHidden, setIsHidden] = React.useState<boolean>(true);

    const toggleSearchbar = () => {
        setIsHidden(!isHidden);
    };

    const ref = React.useRef<HTMLDivElement>(null);
    const [isActive, setIsActive] = React.useState<boolean>(false);

    React.useEffect(() => {
        const checkIfClickedOutside = (e: MouseEvent) => {
            if (
                isActive &&
                ref.current &&
                !ref.current.contains(e.target as Node)
            ) {
                setIsActive(false);
            }
        };

        document.addEventListener("mousedown", checkIfClickedOutside);

        return () => {
            document.removeEventListener("mousedown", checkIfClickedOutside);
        };
    }, [isActive]);

    return (
        <header className="columns header">
            {/* <div data-testid="back-btn" className="icon">
                <BackIcon />
            </div> */}

            <Search hidden={isHidden} onClick={toggleSearchbar} />

            <ShowMore
                dropdownRef={ref}
                state={isActive}
                setState={setIsActive}
            />
        </header>
    );
};
