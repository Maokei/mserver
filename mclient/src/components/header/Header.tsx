import React, { ChangeEvent } from "react";
import { Link } from "react-router-dom";
import { Search } from "./Search";
import { ShowMore } from "./ShowMore";
import styles from "./header.module.scss";
import { BackIcon } from "../../assets/back-vector";

export const Header = () => {
    const [inputTexts, setInputTexts] = React.useState<string>("");

    const handleKeyDown = (e: ChangeEvent<HTMLInputElement>) => {
        console.log(e.target.value);
    };

    const [isHidden, setIsHidden] = React.useState<boolean>(true);

    const toggleSearchbar = () => {
        setIsHidden(!isHidden);
    };

    const ref = React.useRef<HTMLDivElement>(null);
    const [isActive, setIsActive] = React.useState<boolean>(false);

    React.useEffect(() => {
        const checkIfClickedOutside = (e: MouseEvent) => {
            if (
                isActive &&
                ref.current &&
                !ref.current.contains(e.target as Node)
            ) {
                setIsActive(false);
            }
        };

        document.addEventListener("mousedown", checkIfClickedOutside);

        return () => {
            document.removeEventListener("mousedown", checkIfClickedOutside);
        };
    }, [isActive]);

    return (
        <header className={`${styles.header}`}>
            {window.location.pathname === "/" ||
            window.location.pathname === "/library" ? (
                <Search
                    hidden={isHidden}
                    onClick={toggleSearchbar}
                    value={inputTexts}
                    setKeywords={setInputTexts}
                    onKeyDown={handleKeyDown}
                />
            ) : (
                <div
                    data-testid="back-btn"
                    className={`icon ${styles.backBtn}`}
                >
                    <Link to="/">
                        <BackIcon />
                    </Link>
                </div>
            )}

            <ShowMore
                dropdownRef={ref}
                state={isActive}
                setState={setIsActive}
            />
        </header>
    );
};
