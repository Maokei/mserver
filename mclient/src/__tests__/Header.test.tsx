import React from "react";
import {
    getByTestId,
    // cleanup,
    fireEvent,
    render,
    screen,
    getAllByRole,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import { Header } from "../components/Header";
import { ShowMore } from "../components/ShowMore";
import { Search } from "../components/Search";
import userEvent from "@testing-library/user-event";

let container: any = null;

beforeEach(() => {
    container = render(<Header />).container;
});

describe("Page should have a header component", () => {
    it("should show show-more-icon", () => {
        expect(getByTestId(container, "show-more")).toBeTruthy();
    });

    it("should have a dropdown menu", () => {
        expect(getByTestId(container, "dropdown-menu-test")).toBeTruthy();
    });

    it("should show search", () => {
        expect(getByTestId(container, "search")).toBeTruthy();
    });

    it("should have a search button", () => {
        expect(getByTestId(container, "search-btn")).toBeTruthy();
    });

    it("should have a search input", () => {
        expect(getByTestId(container, "search-input")).toBeTruthy();
    });
});

// afterEach(cleanup);

describe("Test input functions", () => {
    it("render input", () => {
        const tree = renderer
            .create(
                <Search
                    value="something"
                    onKeyDown={() => {}}
                    setKeywords={() => {}}
                    onClick={function (): void {
                        throw new Error("Function not implemented.");
                    }}
                    hidden={false}
                />
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("call setKeywords when user type in the input", () => {
        const setValue = jest.fn();

        render(
            <Search
                value=""
                onKeyDown={() => {}}
                setKeywords={setValue}
                onClick={function (): void {
                    throw new Error("Function not implemented.");
                }}
                hidden={false}
            />
        );

        userEvent.type(screen.getByPlaceholderText("search"), "M");
        expect(setValue).toHaveBeenCalledWith("M");
    });
});

describe("Test Button component", () => {
    it("Test click event", () => {
        const mockCallBack = jest.fn();

        const button = shallow(<button onClick={mockCallBack}>Ok!</button>);
        button.find("button").simulate("click");
        expect(mockCallBack.mock.calls.length).toEqual(1);
    });
});

describe("Test dropdown", () => {
    it("runs without crashing", () => {
        render(
            <ShowMore
                dropdownRef={undefined}
                state={false}
                setState={() => {}}
            />
        );
    });

    it("can show that the dropdown has children", () => {
        const dropdown = render(
            <ShowMore
                dropdownRef={undefined}
                state={false}
                setState={() => {}}
            />
        ).container;
        const display = dropdown.children[0];

        expect(display.textContent).toContain("dropdown");
        // console.log(display.textContent);

        fireEvent.click(dropdown);

        const dropdownOptions = getAllByRole(dropdown, "menu");
        fireEvent.click(dropdownOptions[0]);

        expect(display.ELEMENT_NODE).toBe(1);

        // console.log(display.ELEMENT_NODE);
    });
});
