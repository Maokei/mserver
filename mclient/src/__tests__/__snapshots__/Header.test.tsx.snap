// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test input functions render input 1`] = `
<div
  className="control column is-four-fifths search-wrapper"
  data-testid="search"
>
  <button
    className="button icon search-btn"
    data-testid="search-btn"
    onClick={[Function]}
  >
    <svg>
      search-vector.svg
    </svg>
  </button>
  <input
    className="input is-large "
    data-testid="search-input"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="search"
    type="text"
    value="something"
  />
</div>
`;
